### PR TITLE
towards BifurcationKit @0.4

### PR DESCRIPTION
**Pull request by rveltz**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: test/extensions/Project.toml [hunk 1]
```diff
@@ -8,5 +8,6 @@ ModelingToolkit = "961ee093-0014-501f-94e3-6117800e7a78"
 OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
 SciMLSensitivity = "1ed8b502-d754-442c-8d5d-10ac956f44a1"
 SciMLStructures = "53ae85a6-f571-4167-b2af-e1d143709226"
+SetField = "efcf1570-3423-57d1-acb7-fd33fddbac46"
```
Comment (ChrisRackauckas) [inline test/extensions/Project.toml:11]:
```suggestion
Setfield = "efcf1570-3423-57d1-acb7-fd33fddbac46"
```

File: test/extensions/bifurcationkit.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-using BifurcationKit, ModelingToolkit, Test
+using BifurcationKit, ModelingToolkit, SetField, Test
```
Comment (ChrisRackauckas) [inline test/extensions/bifurcationkit.jl:1]:
```suggestion
using BifurcationKit, ModelingToolkit, Setfield, Test
```

File: test/extensions/bifurcationkit.jl [hunk 3]
```diff
@@ -36,7 +36,7 @@ let
     bprob_BK = BifurcationProblem(f_BK,
         [1.0, 1.0],
         [-1.0, 1.0],
-        (@lens _[1]);
+        (SetField.@lens _[1]);
```
Comment (ChrisRackauckas) [inline test/extensions/bifurcationkit.jl:39]:
```suggestion
        (Setfield.@lens _[1]);
```

**Comment by ChrisRackauckas** [comment]
`AssertionError: lens isa Int || lens isa AllOpticTypes` @rveltz do you know what this is?

**Comment by rveltz** [comment]
Yes it is to ensure that the parameter axis is an int or an @optic (from accessors.jl)

**Comment by rveltz** [comment]
Note that you don't need setfield anymore and it is re exported by BK. Not sure why you put it back

**Comment by ChrisRackauckas** [comment]
`  LoadError: UndefVarError: `@lens` not defined in `Main.var"##BifurcationKit Extension Test#294"` I think doing what you are saying leads to this error?

**Comment by rveltz** [comment]
I am confused. In BK@0.4 I replaced Setfield by Accessors. They should not be any `@lens` anywhere in the code

> `AssertionError: lens isa Int || lens isa AllOpticTypes` @rveltz do you know what this is?

Yes. It is to allow to pass `@optic _.bla` or an `Int` as a parameter axis

**Comment by rveltz** [comment]
replace the line https://github.com/SciML/ModelingToolkit.jl/blob/master/test/extensions/bifurcationkit.jl#L39 with `BifurcationKit.@optic _[1]` and it should work

**Comment by ChrisRackauckas** [comment]
```julia
  LoadError: MethodError: no method matching (::Main.var"##BifurcationKit Extension Test#289".var"#2#6")(::Vector{Float64}, ::Float64; iter::BifurcationKit.ContIterable{BifurcationKit.EquilibriumCont, BifurcationKit.BifurcationProblem{BifurcationKit.BifFunction{Main.var"##BifurcationKit Extension Test#289".var"#f_BK#5", BifurcationKit.var"#8#24", Nothing, BifurcationKit.var"#6#22", Nothing, BifurcationKit.var"#11#28"{BifurcationKit.var"#d1Fad#26"}, BifurcationKit.var"#13#30", BifurcationKit.var"#15#32", BifurcationKit.var"#17#34", Bool, Float64}, Vector{Float64}, Vector{Float64}, Accessors.IndexLens{Tuple{Int64}}, typeof(BifurcationKit.plot_default), Main.var"##BifurcationKit Extension Test#289".var"#2#6", typeof(BifurcationKit.save_solution_default)}, BifurcationKit.PALC{BifurcationKit.Secant, BifurcationKit.MatrixBLS{BifurcationKit.DefaultLS}, Float64, BifurcationKit.DotTheta{BifurcationKit.var"#279#281", BifurcationKit.var"#280#282"}}, Float64, BifurcationKit.DefaultLS, BifurcationKit.DefaultEig{typeof(real)}, typeof(LinearAlgebra.norm), typeof(BifurcationKit.finalise_default), typeof(BifurcationKit.cb_default), Nothing}, state::BifurcationKit.ContState{BifurcationKit.BorderedArray{Vector{Float64}, Float64}, Float64, Vector{ComplexF64}, Matrix{ComplexF64}, Tuple{Nothing, Nothing}})
  
  Closest candidates are:
    (::Main.var"##BifurcationKit Extension Test#289".var"#2#6")(::Any, ::Any) got unsupported keyword arguments "iter", "state"
     @ Main.var"##BifurcationKit Extension Test#289" ~/work/ModelingToolkit.jl/ModelingToolkit.jl/test/extensions/bifurcationkit.jl:40
```

Is this from some anonymous function generated by optic? I have no idea what's going on here or what `@optic` even is: is there a simpler way to write this?

**Comment by rveltz** [comment]
>  I have no idea what's going on here or what @optic even is: is there a simpler way to write this?

An optic is from Accessors.jl
It allows to specify a parameter axis. For example, if you pass a named typle as a parameter

```
pars = (a=1., b=2.)
```

and would like to continue the solutions wrt to a, you can pass `@optic _.a`. If you pass an array and want to continue wrt to the first parameter, you can pass `@optic _[1]`. In this last case, you could pass directly `1` to `continuation`. See also https://github.com/bifurcationkit/BifurcationKit.jl/issues/159

File: test/extensions/bifurcationkit.jl [hunk 4]
```diff
@@ -36,7 +36,7 @@ let
     bprob_BK = BifurcationProblem(f_BK,
         [1.0, 1.0],
         [-1.0, 1.0],
-        (@lens _[1]);
+        (BifurcationKit.@optic _[1]);
         record_from_solution = (x, p) -> x[1])
```
Comment (ChrisRackauckas) [inline test/extensions/bifurcationkit.jl:40]:
```suggestion
        record_from_solution = (x, p; k...) -> x[1])
```

**Comment by ChrisRackauckas** [comment]
Awesome, thanks so much!

**Comment by rveltz** [comment]
No problem. I am embarassed, I thought my PR would have been more straightforward

**Pull request merged**